allennlp/nn/util.py

```diff
@@ -192,8 +192,8 @@ def masked_softmax(vector, mask):
         result = torch.nn.functional.softmax(vector, dim=-1)
     else:
         # To limit numerical errors from large vector elements outside the mask, we zero these out.
-        result = torch.nn.functional.softmax(vector * mask, dim=-1)
-        result = result * mask
+        result = torch.nn.functional.softmax(replace_masked_values(vector, mask, 0.0), dim=-1)
```

Comment:
Did you verify that this actually reduces memory usage?  I think this actually _adds_ an operation, because `replace_masked_values` does a subtraction, then a `masked_fill()`.

Reply:
Before changing this, I get an OOM error at 
`mask = mask.expand_as(tensor).contiguous().float()`, whereas after I get it in the matrix multiplication afterward.

Reply:
Something to consider, most of times we don't need to expand mask if it is broadcastable with tensor.

Reply:
Right - unfortunately we have many cases where they aren't, such as in the multi-headed attention.

Reply:
We actually _can_ remove that expand, as an `unsqueeze()` is enough.  #1687 removes the `last_dim_applicator` entirely, which should further help your memory problems.